File: classes/CrackupFile.php
<?php
/**
 * CrackupFile represents a filesystem file.
 *
 * @author Ryan Grove <ryan@wonko.com>
 * @package Crackup
 */
class CrackupFile extends CrackupFileSystemObject {
  // -- Private Instance Variables ---------------------------------------------  
  private $_fileHash = '';
  
  // -- Constructor ------------------------------------------------------------
  
  /**
   * Constructs a new CrackupFile object.
   *
   * @param String $filename filename
   */
  public function __construct($filename) {
    if (!is_file($filename)) {
      throw new Exception($filename.' does not exist or is not a file');
    }
    
    $this->_fileHash = hash_file('sha256', $filename);
    
    parent::__construct($filename);
  }
  
  // -- Public Instance Methods ------------------------------------------------
  
  /**
   * Gets the SHA256 hash value of the file represented by this CrackupFile
   * instance as of the moment the class was instantiated.
   *
   * @return String
   */
  public function getFileHash() {
    return $this->_fileHash;
  }
  
  /**
   * Restores the remote copy of this file to the specified local path.
   * 
   * @param String $localPath
   */
  public function restore($localPath) {
    $localPath = rtrim($localPath, '/').'/'.str_replace(':', '', dirname(
        $this->getName()));
    $localFile  = $localPath.'/'.basename($this->getName());
    $remoteFile = Crackup::$remote.'/crackup_'.$this->getNameHash();

    // In Windows, PHP's mkdir() function only works with backslashes because
    // PHP was written by fucktards from the planet Spengo.
    $windowsLocalPath = str_replace("/", "\\", $localPath);
    $isWindows = strpos(php_uname('s'), 'Windows') === 0;
    
    Crackup::debug($localFile);

    if (!is_dir($localPath)) {
      if (!@mkdir(($isWindows ? $windowsLocalPath : $localPath), 0750, true)) {
        Crackup::error('Unable to create local directory: "'.$localPath.'"');
      }
    }
    
    // Download the file.
    $tmpFile = Crackup::getTempFile();
    
    try {
      Crackup::$driver->get($remoteFile, $tmpFile);
    }
    catch (Exception $e) {
      @unlink($tmpFile);
      Crackup::error('Unable to restore file: '.$localFile);
    }
    
    if (defined('CRACKUP_NOGPG')) {
      Crackup::decompressFile($tmpFile, $localFile);
    }
    else {
      Crackup::decryptFile($tmpFile, $localFile);
    }
    
    @unlink($tmpFile);
  }
}
?>

File: crackup/lib/CrackupFile.rb
require 'digest/sha2'
require 'fileutils'

module Crackup

  # Represents a file on the local filesystem.
  class CrackupFile < CrackupFileSystemObject
    attr_reader :file_hash, :url
  
    def initialize(filename)
      unless File.file?(filename)
        raise ArgumentError, "#{filename} is not a file"
      end
      
      super(filename)
      
      # Get the file's SHA256 hash.
      digest = Digest::SHA256.new()
      
      File.open(filename, 'rb') do |file|
        until file.eof? do
          digest << file.read(1048576)
        end
      end

      @file_hash = digest.hexdigest()
      @url       = "#{Crackup::driver.url}/crackup_#{@name_hash}"
    end
    
    # Removes this file from the remote location.
    def remove
      Crackup::debug '--> ' + @name
      Crackup::driver.delete(@url)
      
    rescue => e
      Crackup::error "Unable to delete remote file: #{@url}"
    end

    # Restores the remote copy of this file to the local path specified by
    # <em>path</em>.
    def restore(path)
      path     = path.chomp('/') + '/' + File.dirname(@name).delete(':')
      filename = path + '/' + File.basename(@name)

      Crackup::debug '--> ' + filename
      
      # Create the path if it doesn't exist.
      unless File.directory?(path)
        begin
          FileUtils.mkdir_p(path)
        rescue => e
          Crackup::error "Unable to create local directory: #{path}"
        end
      end
      
      # Download the remote file.
      tempfile = Crackup::get_tempfile()
      
      begin
        Crackup::driver.get(@url, tempfile)
      rescue => e
        Crackup::error "Unable to restore file: #{filename}"
      end
      
      # Decompress/decrypt the file.
      if Crackup::options[:passphrase].nil?
        Crackup::decompress_file(tempfile, filename)
      else
        Crackup::decrypt_file(tempfile, filename)
      end
    end
    
    # Uploads this file to the remote location.
    def update
      Crackup::debug '--> ' + @name
      
      # Compress/encrypt the file.
      tempfile = Crackup::get_tempfile()
      
      if Crackup::options[:passphrase].nil?
        Crackup::compress_file(@name, tempfile)
      else
        Crackup::encrypt_file(@name, tempfile)
      end
      
      # Upload the file.
      begin
        Crackup::driver.put(@url, tempfile)
      rescue => e
        Crackup::error "Unable to upload file: #{@name}"
      end
    end
  end
end
